infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/limits/ram/ContainerResourceLimitRequestProvisioner.java

```diff
@@ -0,0 +1,117 @@
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.Containers;
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.KubernetesSize;
+
+/**
+ * Sets or overrides Kubernetes container RAM and CPU limit and request if corresponding attributes
+ * are present in machine corresponding to the container.
+ *
+ * <p>There are two memory-related properties: - che.workspace.default_memory_limit_mb - defines
+ * default machine memory limit - che.workspace.default_memory_request_mb - defines default
+ * requested machine memory allocation. Similarly, two CPU-related properties: -
+ * che.workspace.default_cpu_limit_cores - defines default machine CPU limit and -
+ * che.workspace.default_cpu_request_cores - defines default machine CPU request
+ *
+ * @author Anton Korneta
+ * @auhtor Max Shaposhnyk
+ */
+@Singleton
+public class ContainerResourceLimitRequestProvisioner implements ConfigurationProvisioner {
+
+  private final ResourceLimitAttributesProvisioner resourceLimitAttributesProvisioner;
+
+  private final long defaultMachineMaxMemorySizeAttribute;
+  private final long defaultMachineRequestMemorySizeAttribute;
+  private final float defaultMachineCpuLimitAttribute;
+  private final float defaultMachineCpuRequestAttribute;
+
+  @Inject
+  public ContainerResourceLimitRequestProvisioner(
+      @Named("che.workspace.default_memory_limit_mb") long defaultMachineMaxMemorySizeAttribute,
+      @Named("che.workspace.default_memory_request_mb")
+          long defaultMachineRequestMemorySizeAttribute,
+      @Named("che.workspace.default_cpu_limit_cores") String defaultMachineCpuLimitAttribute,
+      @Named("che.workspace.default_cpu_request_cores") String defaultMachineCpuRequestAttribute,
+      ResourceLimitAttributesProvisioner resourceLimitAttributesProvisioner) {
+    this.resourceLimitAttributesProvisioner = resourceLimitAttributesProvisioner;
+    this.defaultMachineMaxMemorySizeAttribute = defaultMachineMaxMemorySizeAttribute * 1024 * 1024;
+    this.defaultMachineRequestMemorySizeAttribute =
+        defaultMachineRequestMemorySizeAttribute * 1024 * 1024;
+    this.defaultMachineCpuLimitAttribute = KubernetesSize.toCores(defaultMachineCpuLimitAttribute);
+    this.defaultMachineCpuRequestAttribute =
+        KubernetesSize.toCores(defaultMachineCpuRequestAttribute);
+  }
+
+  @Override
+  @Traced
+  public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+
+    TracingTags.WORKSPACE_ID.set(identity::getWorkspaceId);
+
+    final Map<String, InternalMachineConfig> machines = k8sEnv.getMachines();
+    for (PodData pod : k8sEnv.getPodsData().values()) {
+      for (Container container : pod.getSpec().getContainers()) {
+
+        // make sure that machine configs have settings for RAM limit and request
+        InternalMachineConfig machineConfig = machines.get(machineName(pod, container));
+        resourceLimitAttributesProvisioner.provisionMemory(
```

Comment:
So this method basically does some sanity checks on the provided limits and applies the sanitized values to the map. You then get these values again from the map in the code below to apply these values elsewhere.

What if you just returned the sanitized values and applied them here to all the places that need it?

Reply:
it incapsulates some logic = comptures correct limit, that sets values into map only if they're not set before. So imo it's correct to have it separately in the same well-testes method instead of using on-demand in several places.